### DISCUSSION TITLE
Uptime Monitoring: Share your experience and expectations

### DISCUSSION THREAD

**Discussion by therealarkin** [getsentry/sentry#51110] [Problem Understanding]
Hello Everyone!

I'm Eran from the Sentry product team. We're always looking for opportunities to improve and extend our services to help you manage your applications more effectively. One such area that has come to our attention is Uptime Monitoring!

As we contemplate a service in this direction, we'd love to understand how you currently handle uptime monitoring, what challenges you face, and how an integrated Sentry service might add value. 

Here are some starter questions to spur the conversation:
1. How do you currently monitor your websites or web applications' uptime? Which tools or services do you use for this purpose, if any?
2. How effective are these tools? What aspects do you like, and where do you feel they fall short?
3. How often do you encounter website downtime? What are the usual causes (e.g., server issues, code bugs, etc.)?
4. What features would you like to see in an uptime monitoring service from Sentry?

Your input is invaluable to us as we explore this new venture. All thoughts, experiences, and ideas are welcome! Even if you have never used an uptime monitoring service, we're curious to hear what you'd expect from one.

Thanks for taking the time to contribute. We're looking forward to hearing your thoughts and ideas!

**Comment by defp** [comment]
I like https://gatus.io/ feature rich, easy customization

**Reply by therealarkin** [reply]
@defp - thanks! Which features have you particularly found helpful and why?

**Comment by cleptric** [comment]
1. https://betteruptime.com/
2. Super effective. They offer a variety of notification channels, status pages, and a deep level of customization of your uptime checks.
3. Mostly borked deploys.
4. Multi-region support (check uptime from various pops around the globe), a simple UI, an API to automate uptime check creation and SSL certificate monitoring.

**Reply by therealarkin** [reply]
Thanks @cleptric! Which customization you found the most useful and why?
Also, do you mean broken deploys?

**Reply by cleptric** [reply]
Yeah, borked is slang for not working properly 😬

![Screenshot 2023-06-21 at 10 41 01](https://github.com/getsentry/sentry/assets/6617432/311938d9-22f0-4941-bead-8e8cd0e33c01)

I like all the different assertions a monitor can have when checking for uptime.

 
![Screenshot 2023-06-21 at 10 40 08](https://github.com/getsentry/sentry/assets/6617432/f3bbae3f-93e8-4164-a078-4f6012a8847f)

For the checks, there are a lot of parameters someone can tweak, like the check interval, adding auth and payloads, following redirects or not, when to check for expired SSL certs and much more.

**Reply by aldy505** [reply]
+1 for SSL certificate monitoring, most free uptime monitoring software don't have a good emphasis on that one.

**Comment by PaitoAnderson** [comment]
1. https://uptimerobot.com/
2. No issues in the 6+ years using it, quite basic but does the job.
3. ~4 a year, usually odd web server crash or cloud outage.
4. Support to call multiple people from a known phone number so we can set this as an emergency contact in our phone.

**Comment by estahn** [comment]
We started using Grafana Synthetic Monitoring quite recently.

https://grafana.com/blog/2022/03/10/best-practices-for-alerting-on-synthetic-monitoring-metrics-in-grafana-cloud/

**Reply by therealarkin** [reply]
Thank you for sharing! What do you like about it? What do you dislike?

**Comment by aldy505** [comment]
1. https://github.com/louislam/uptime-kuma
2. I like it because it's free. Can monitor lots of types (HTTP, ICMP, DNS, gRPC request, MQTT, MariaDB, Postgres, Push (like the reverse of webhook), and even Steam game server). Has TLS certificate expiry notification and status
3. Mostly because of network failures (ISP uptime SLA problems), and sometimes because of it's in the middle of a release pipeline.
4. Everything that Uptime Kuma has + a separate UI that can run on a separate network (like literally I can spawn a VM for the dashboard on another cloud provider). See https://github.com/louislam/uptime-kuma/issues/3385 for detail on how I use this.

**Comment by dwd** [comment]
I have used:
* SmokePing (pretty good basic visualisation, but no alerting)
* New Relic (amazingly good, but gets pricey very quickly if you take full advantage)
* Google Cloud's own (fair, but limited, and obviously can't tell you if the entire data centre is underwater/melted/etc).
* A bunch of other stuff (like, I once used a cronjob that drove a shell script that drove ping and SNPP, and if you know what the latter even is then we should get together and cry into a beer together).

Things I like:
* New Relic's integration was really nice - so if a response was slower than expected, we could dive into the APM of the tested API and see why.
* New Relic did all the TLS cert validation, too, but would also warn if the expiry was less than (say) a week away.
* NR's full synthetics stuff is impressive - it'll run a whole headless browser for you if you want, as well as a JS script of Stuff To Do.
* SmokePing's visualisation are great for seeing beyond "Is this thing up?" - the display of distribution on the y axis is really nice. Ping me, and I'll give you a link to my personal one.
* NR's "where would you like this to be run from" is pretty nice.
* SmokePing can do ICMP and DNS as well (actually better than) other probes.

You might notice I don't say anything much positive about Google's - it gives value, for sure, and it does the basic job. But beyond dumping an alert out if we've a long running database migration on deploy, it's not doing much more.

So my future plan is this: https://github.com/dwd/sentrynet - which is (or will be, it's only doing very basic stuff now) an attempt to make a synthetics source we can deploy mega-cheap in a bunch of locations (hosting a small docker image is like $5 a month these days), and give us the shiny integration with Sentry (Issues and Performance) that my tiny heart craves.

So... your questions:

1. Yeah, Google Cloud.
2. I mean, it detects downtime, but absolutely nothing beyond that.
3. We get a ~false positive if a migration takes too long. Other than that, our longest downtime was because Google's data centre melted.
4. Everything! Mostly, I think I can Sentrynet to get the right data in, but the visualisations aren't as shiny as I'd like. But downtime is (or should be) rare - I want to be able to find brown-outs in service more effectively, and that means more flexible alerting and reporting than most of these tools offer. New Relic is an exception here, but I think Sentrynet + Sentry Anomalies + some better visualiation could be a winner.

**Comment by ipeterov** [comment]
What I find really useful is an ability to create Selenium-like tests, and not just checking that a URL returns 200 to some response. Uptime.com [has this feature](https://uptime.com/synthetic-monitoring).

It's really useful for downtimes caused by interaction of frontend & backend (for example, backend using an incorrect CDN url, resulting in a blank page in production). Those are really hard to cover by regular tests, so having a tool notify you immediately after you broke something, instead of waiting for users to wake you up is really nice.

The thing I disliked about the uptime.com implementation is that it runs at most every 5 minutes, which makes for a not-very-accurate uptime calculation, and also delays my average time to reaction by 2.5 minutes, which is a pretty big deal. It would be nice to see a once-a-minute option.

**Comment by edgariscoding** [comment]
Definitely interested in this and it seems like a perfect feature for Sentry.

**Comment by darioseidl** [comment]
We've been using [UptimeRobot](https://uptimerobot.com/) and [HetrixTools](https://hetrixtools.com/), free or paid, depending on the company. Both work quite well.

Features that I like/need: 

- checking for TLS certificate expiration
- checking pages with basic auth
- monitoring other services/ports, not just HTTP(S)
- blacklist monitoring
- generating a (public) status page for all monitors
- sending notifications to mail/Slack/Telegram channels (so that you don't need yet another app on your phone)

**Comment by JoseAlban** [comment]
Using [freshping.io](https://www.freshworks.com/) ATM

Pros
- Very basic, it tests pre-set status endpoints, including CDN and servers
- It summarises uptime and sends email notifications with percentage
- public status page for all monitors, with optional custom domain

Cons
- Does not wait for canvas to be fully rendered, and our app is canvas-heavy
- Does not have any advanced feature specific checks after the single page app is loaded, so we can be up but unavailable feature wise

**Comment by orenaksakal** [comment]
- How do you currently monitor your websites or web applications' uptime? Which tools or services do you use for this purpose, if any?
**Self hosted uptimekuma**

- How effective are these tools? What aspects do you like, and where do you feel they fall short?
**I like it but would love to have uptime/error reporting and performance metrics in one place like glitchtip**

- How often do you encounter website downtime? What are the usual causes (e.g., server issues, code bugs, etc.)?
**Not very often**

- What features would you like to see in an uptime monitoring service from Sentry?
**Ping, expected response body, 1m, 5m etc incremental checks, SSL checks and notifications, monitoring grouping e.g. project-api, project-site, response timeout, simple status pages**

**Comment by adetch** [comment]
@therealarkin any news on an offering in this space? If nothing in sentry, it would be great to just have an integration w/ https://uptime.betterstack.com/ to highlight periods with downtime

**Comment by d3thomas** [comment]
@therealarkin What is the status on this? Currently looking for a monitoring solution for my FastAPI's. All looks good and works great on the implementation. The Uptime Monitoring is not yet there I see, but is a requirement for a client of mine. Would love to have that in Sentry! Can you tell us a bit more on if this is still on the roadmap, if yes when can we expect it?

**Reply by gaprl** [reply]
Hey there @d3thomas, we're happy to announce we've recently released an [early access version of uptime](https://changelog.sentry.dev/changelog/uptime-monitoring-early-adopters-release/)! We'd love for you to check it out and provide  any feedback at uptime-feedback@sentry.io.

**Reply by d3thomas** [reply]
@gaprl Great thank you, I will check it out!

**Comment by edgariscoding** [comment]
Also interested in this

**Comment by tdtm** [comment]
We're currently using BetterUptime so fwiw my key points are similar to others comparing with BetterUptime.

A key thing I would add, and that Sentry is in a perfect position to bring, would be to have more levels on public status pages (or internal dashboards for that matter). So rather than up/down, we'd also have "degraded" if errors are above a certain reference level. 

Even better yet, having this grouped by certain rules instead of just being based on a certain page returning a certain status code and/or string. I would imagine the main rule would be route segments for most users but I suppose that can be a bit arbitrary.

For example, Sentry would know that I have a high error rate at example.com/dashboard but everything else is OK; so it could be "Up - Dashboard degraded". A high error rate at example.com/members/:memberId/edit could be "Up - Member management degraded". But then very high error rates thorough the whole app could just be classified as "down".

This would be much smarter than the oversimplified up/down that most services currently allow.

**Reply by tdtm** [reply]
Regarding the degraded aspect, I love the idea of responding to the uptime monitor with a standardized response like proposed in https://datatracker.ietf.org/doc/html/draft-inadarei-api-health-check so the uptime monitor can flag as....

Downtime if:
- no response (eg network issue)
- error response (5xx)
- monitor reports "fail"

Or degraded if:
- response time is X% above baseline
- getting intermittent errors (say the threshold is 5 failures in a row, and we did not get 5 failures in a row but we got 50% failures in the last 5 minutes)
- monitor reports "warn" (to use the RFC language)
- Sentry's error tracker observes significantly more errors than normal (e.g. >10% error rate)

Or just up/operating normally otherwise.

We'd also prefer putting this behind some sort of simple authentication to avoid disclosing unnecessary details much to a potential  attacker.

**Reply by tdtm** [reply]
For a multi-service setup it would also depend on what "area" of the application is encountering problems. 

The authentication being down means the whole product is down, as it's critical. Nobody can do anything is auth isn't working.

If the only thing failing is file uploads, this means the product experience is degraded, but I woudn't report that as downtime.

**Comment by therealarkin** [comment]
This is 🔥 congrats to the team: https://sentry.io/changelog/uptime-monitoring-early-adopters-release. Looks like @gaprl will take it from here 😃

**Comment by vpmedia** [comment]
Hi!
I would like an option in the Sentry UI to completely opt-out from this feature.
I have no auto uptime alert configured and robots.txt deny all is in place, still I'm getting monitoring calls to my servers.

Thanks!

**Comment by vishal2-wiai** [comment]
Is dedicated Uptime Monitoring available in the open source self-hosted version as well? I am unable to find a way to enable it.

**Reply by aldy505** [reply]
Yes it does! See here https://github.com/getsentry/self-hosted/issues/3304

**Comment by fnagel** [comment]
I would need response time monitoring. Having stats over the response time is a good metric if something is wrong.

**Comment by yashshanker** [comment]
How can I configure conditions/notifications for Uptime Alerts ?

The screenshots in the [docs](https://docs.sentry.io/product/alerts/uptime-monitoring/) show Set Conditions section, which is not there under Uptime Alert.

**Comment by noqcks** [comment]
Just trying uptime monitoring for the first time today and have some thoughts. 

1. What status codes will cause uptime monitoring to fail? There's nothing that indicates what status codes will trigger it. I would hope and assume that anything other than 2xx fails.

<img width="1257" alt="image" src="https://github.com/user-attachments/assets/e5abcc13-5766-444e-a816-ebb3dfb9bd14" />

2. What is ownership? I have no idea where my alerts will go if an uptime monitor fails. The text describing ownership also doesnt tell me how alerts are routed. 

3. How do I test it? Is there a way to trigger an alert so that I can see what the failure message looks like in Slack? Would love that.

**Reply by venkigvs2002** [reply]
got any reply??

**Reply by gaprl** [reply]
1. We use 2xx status codes, you can read more about it in our [uptime check criteria](https://docs.sentry.io/product/uptime-monitoring/#uptime-check-criteria). We do plan on enabling you to customize this setting.
2. Ownership is used to define the assignee of a downtime issue when we detect an outage for the configured URL. Here's some more detail on [how notifications for uptime works](https://docs.sentry.io/product/uptime-monitoring/#notifications).
3. We don't offer a way to test your check currently, but that's definitely something we're looking into building!

**Comment by philprime** [comment]
We started using Sentry Uptime Monitoring for an instance of HashiCorp Vault polling the `/sys/health` endpoint, which returns different status codes but also details in the JSON response body ([Documentation](https://developer.hashicorp.com/vault/api-docs/system/health)).

To summarize the issue:
- Vault allows storing secrets and can be sealed from access until administrator interaction.
- If the vault is sealed, the endpoint will return a 503 HTTP Status
- If the deployment crashes, the load balancer in my cluster also returns 503
- Therefore status code is not sufficient to identify the problem

What I need:
- Check fields in the response JSON and use it as uptime conditions, i.e. `'.sealed' IS 'false'`
```json
{
  "clock_skew_ms": 0,
  "cluster_id": "995f749a-9bc3-fbe0-a185-ccba6999fc73",
  "cluster_name": "vault-cluster-e65c0563",
  "echo_duration_ms": 0,
  "enterprise": false,
  "initialized": true,
  "performance_standby": false,
  "replication_dr_mode": "disabled",
  "replication_performance_mode": "disabled",
  "sealed": false,
  "server_time_utc": 1709559327,
  "standby": false,
  "version": "1.16.0-rc2"
}
```
- [JQ](https://jqlang.org/tutorial/) could be a reference for key-path-based access

**Reply by tdtm** [reply]
Agreed, this would also be useful for RFC 2629 health checks, as well as any similar custom schema, so a generic jq-style feature would be very versatile

**Reply by christianrr** [reply]
Indeed, checking the response for some specific content is super important for uptime monitors IMO.
JSON-support would be great, however, checking the string content of the response with contains/not-contains would be sufficient for most cases (maybe this would be an easier first step)

**Comment by brianguertin** [comment]
It's nice to see uptime monitoring in Sentry. But it would be nice if there was a public "status" page? Like https://status.sentry.io/

**Reply by edgariscoding** [reply]
What would be amazing!

**Comment by raheelkhan** [comment]
Not sure if I missed something, I have created a team on Sentry `devops` and on the Uptime Monitor Alert, I have set the ownership to the `#devops` team. When the target is down, The team members get email but no slack messages. 

We have already integrated Slack with Sentry and on other channels we get the issues but in the Uptime Alert Monitor there is no way to set Slack as the notification mechanism. It just ask to select the Ownership.

On Slack side, in the public channel `#devops-alerts` I have ran commands such as `/sentry link team devops` and I get the message 

> The devops team has already been linked to a Slack channel.

Is this feature not implemented yet for Uptime Monitors ?

**Reply by andinger** [reply]
i've had this issue too ... had to disable the personal issue workflow notifications (or the owner team notifications) and created an issue alert which itself sends the alerts to the slack channel. you can filter the issue alerts to `category=outage`. uptime monitor simply creates issues in the project feed ... works also in self-hosted

**Comment by JackLeo** [comment]
This feature is nice to have, but really lacks better escalation support. The main benefit of other platforms is that they will call me in the middle of the night if they're set up and I won't acknowledge a downtime.

**Comment by JEONSEWON** [comment]
The response body validation point is spot on. Status codes alone miss a whole class of failures — the endpoint is "up" but returning an error payload, or a critical field has an unexpected value.

I ran into the same gap and ended up building CheckAPI (checkapi.io) specifically for this. It supports keyword must-be-present / must-be-absent checks on the response body, which covers the HashiCorp Vault case and similar patterns without needing full JQ.

Not a replacement for Sentry's deeper integration, but useful if you need that check today while waiting for this feature to land.

**Comment by andrianvaleanu** [comment]
Integrating uptime into Sentry makes sense since jumping between error logs and status pages is annoying when things break. Most services I use are either too basic or get expensive fast once you add status pages and multi-region checks. Pulsetic is what I've been on lately because the alerts are reliable, though it's another dashboard to manage. 

Here is website: https://pulsetic.com/ if you want to see how they handle it. Having those heartbeat checks right next to my Sentry issue stream would save a lot of time during outages.

**Reply by shermer610-jpg** [reply]
@andrianvaleanu Your point about jumping between error logs and status pages is exactly what we’re hearing from a lot of DevOps leads right now. You clearly have a specific workflow for these outages.

I’m building out an Expert Buying Guide for Sentry on PeerSpot to help other engineers compare these "all-in-one" vs. "best-of-breed" setups (like your Pulsetic use case).

If you're open to it, you can share your experience directly here: (https://www.peerspot.com/interviews/new?product_id=30820).

**Comment by darioandyoshi-tech** [comment]
Great thread! 👋

We built **PulseWatch** (https://pulsewatch.us) for exactly this gap you're describing — specifically the "vendor outage" blind spot.

**What we monitor:**
- 1500+ third-party vendors (Stripe, AWS, Anthropic, Cloudflare, etc.)
- 60-second checks from multiple regions
- Catches outages before status pages update (vendor status pages are often delayed 15-60 min)

**Why we built it:**
Our team kept getting burned by "dependency outages" that our internal monitoring missed. We'd see errors spike, but our uptime tools said "all green" because they only checked *our* endpoints. The actual problem was Stripe/AWS/Anthropic having issues.

**What's different:**
- We don't replace internal monitoring — we complement it
- Focus purely on external dependencies you don't control
- Free tier covers most startups; paid tiers for higher frequency

**Your SentryNet idea sounds solid.** The integration with Sentry Issues would be clutch — correlating your error spikes with vendor outages automatically.

Happy to share what we've learned about multi-region check design or false-positive reduction if useful. Keep building! 🚀

**Comment by darioandyoshi-tech** [comment]
+1 on this! We hit the same need recently.

One thing we've learned: vendor status pages (Stripe, AWS, etc.) are often delayed 15-60 minutes during incidents. So even if Sentry integrates with Statuspage.io, you're still getting late alerts.

We built PulseWatch (https://pulsewatch.us) to catch these earlier by monitoring the actual vendor endpoints directly, not their status pages. Catches outages 10-30 min faster in our testing.

Would love to see Sentry integrate with both approaches — Statuspage for official comms, PulseWatch for early detection. They complement each other well.
